skfem/utils.py

```diff
@@ -307,39 +306,47 @@ def enforce(A: spmatrix,
     D
         Specify either this or ``I``: The set of degree-of-freedom indices to
         enforce (rows/diagonal set to zero/one).
+    overwrite
+        Optionally, the original system is both modified (for performance) and
+        returned (for compatibility with :func:`skfem.utils.solve`).  By
+        default, ``False``.
 
     Returns
     -------
-    EnforcedSystem
+    LinearSystem
         A linear system with the enforced rows/diagonals set to zero/one.
 
     """
     b, x, I, D = _init_bc(A, b, x, I, D)
 
+    Aout = A if overwrite else A.copy()
+    bout: Optional[Union[ndarray, spmatrix]] =\
+        None if b is None or overwrite else b.copy()
+
     # set rows on lhs to zero
-    start = A.indptr[D]
-    stop = A.indptr[D + 1]
+    start = Aout.indptr[D]
+    stop = Aout.indptr[D + 1]
     count = stop - start
     idx = np.ones(count.sum(), dtype=np.int64)
     idx[np.cumsum(count)[:-1]] -= count[:-1]
     idx = np.repeat(start, count) + np.cumsum(idx) - 1
-    A.data[idx] = 0.
+    Aout.data[idx] = 0.
 
     # set diagonal value
-    d = A.diagonal()
+    d = Aout.diagonal()
     d[D] = diag
-    A.setdiag(d)
+    Aout.setdiag(d)
 
-    if b is not None:
-        if isinstance(b, spmatrix):
+    if bout is not None:
+        if isinstance(bout, spmatrix):
             # eigenvalue problem
-            b = enforce(b, D=D, diag=0.)
+            bout = enforce(bout, D=D, diag=0.)
```

Comment:
Should `overwrite=overwrite` be passed on the recursive call?

Reply:
Actually no, there's something else.  On stepping through this with pencil and paper, for `b: spmatrix` I see:
* `overwrite=False`: `bout` is a fresh copy of `b` so actually could well be overwritten but the recursive call has `overwrite=False` and so a second copy of `b` is made in `Aout = A if overwrite else A.copy()`.  That doesn't seem right.  Well, it's not wrong but making a second copy seems wasteful.
* `overwrite=True`: `bout = None` so `bout is not None` fails and only `Aout` is returned.  That's wrong; we do want to return a mass matrix with the Dirichlet condition enforced, i.e. with the Dirichlet rows zeroed.

Reply:
*Aside:&mdash;* I wonder whether it would be clearer to separate the two uses of `b`, instead of having it stand for either right-hand side vector or mass matrix.  Perhaps adding the argument:
```python
M: Optional[spmatrix] = None
```
for the mass matrix, as in [`scipy.sparse.linalg.eigs`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.eigs.html#scipy.sparse.linalg.eigs).

This might be a bit of an overhaul though; I haven't thought it through.

Note that a case with `spmatrix` `b` (or `M`) is not necessarily a generalized eigenvalue problem, it might be an initial value problem #531 , like ex19, and in that case we might have both a mass matrix and a nonzero right-hand side vector.  I think that something like that was asked for in #529, like a heat equation with a source term.